diff --git a/index.d.ts b/index.d.ts
index 43ba6b7fe458e77d152fe0b2f7afeac05d8fc563..979c61d461cc22711033ec43b9d7b749fa4a4396 100755
--- a/index.d.ts
+++ b/index.d.ts
@@ -7,12 +7,11 @@
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // Minimum TypeScript Version: 4.3
 
-/// <reference types="jest" />
-
-import { TestingLibraryMatchers } from './matchers';
+import { TestingLibraryMatchers } from "./matchers";
 
 declare global {
-    namespace jest {
-        interface Matchers<R = void, T = {}> extends TestingLibraryMatchers<typeof expect.stringContaining, R> {}
-    }
+  namespace jest {
+    interface Matchers<R = void, T = {}>
+      extends TestingLibraryMatchers<typeof expect.stringContaining, R> {}
+  }
 }